src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,375 @@
+import org.prebid.server.proto.openrtb.ext.request.ExtImpPrebid;
+import org.prebid.server.proto.openrtb.ext.request.ExtRequest;
+import org.prebid.server.proto.openrtb.ext.request.ExtRequestPrebid;
+import org.prebid.server.proto.openrtb.ext.request.ExtStoredRequest;
+import org.prebid.server.proto.openrtb.ext.response.seatnonbid.NonBid;
+import org.prebid.server.proto.openrtb.ext.response.seatnonbid.SeatNonBid;
+import org.prebid.server.util.HttpUtil;
+import org.prebid.server.version.PrebidVersionProvider;
+import org.prebid.server.vertx.httpclient.HttpClient;
+import org.prebid.server.vertx.httpclient.model.HttpClientResponse;
+
+import java.time.Clock;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GreenbidsAnalyticsReporter implements AnalyticsReporter {
+
+    private static final String PUBLISHER_ID_DYNAMIC_PARAM = "pbuid";
+    private static final String GREENBIDS_SAMPLING_DYNAMIC_PARAM = "greenbidsSampling";
+    private static final String BID_REQUEST_ANALYTICS_EXTENSION_NAME = "greenbids";
+    private static final int RANGE_16_BIT_INTEGER_DIVISION_BASIS = 0x10000;
+    private static final Logger logger = LoggerFactory.getLogger(GreenbidsAnalyticsReporter.class);
+    private final GreenbidsAnalyticsProperties greenbidsAnalyticsProperties;
+    private final JacksonMapper jacksonMapper;
+    private final HttpClient httpClient;
+    private final Clock clock;
+    private final PrebidVersionProvider prebidVersionProvider;
+
+    public GreenbidsAnalyticsReporter(
+            GreenbidsAnalyticsProperties greenbidsAnalyticsProperties,
+            JacksonMapper jacksonMapper,
+            HttpClient httpClient,
+            Clock clock,
+            PrebidVersionProvider prebidVersionProvider) {
+        this.greenbidsAnalyticsProperties = Objects.requireNonNull(greenbidsAnalyticsProperties);
+        this.jacksonMapper = Objects.requireNonNull(jacksonMapper);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.clock = Objects.requireNonNull(clock);
+        this.prebidVersionProvider = Objects.requireNonNull(prebidVersionProvider);
+    }
+
+    @Override
+    public <T> Future<Void> processEvent(T event) {
+        final AuctionContext auctionContext;
+        final BidResponse bidResponse;
+
+        if (event instanceof AmpEvent ampEvent) {
+            auctionContext = ampEvent.getAuctionContext();
+            bidResponse = ampEvent.getBidResponse();
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            auctionContext = auctionEvent.getAuctionContext();
+            bidResponse = auctionEvent.getBidResponse();
+        } else {
+            return Future.failedFuture(new PreBidException("Unprocessable event: " + event.getClass().getName()));
```

Comment:
usually it's not a good practice to expose any internal stuff like class names to the public

so please change the message and if you'd like to have this info it's worth logging it